### PR TITLE
feat(core): add getPeers and peerDiscovery handlers to session actor

### LINKED ISSUE #92: Add "GetPeers" handler to Session actor
#### Context
- Each `Session` actor should be capable of handling `GetPeers` messages (normally coming from the `Anycast` handler of `SessionManager`).
- These `GetPeers` messages are actor messages, not protocol messages (`SessionsManager` should never construct protocol messages, and `Session` should construct them but never handle them from the outside).
- Inside this `GetPeers` handler, a `GetPeers` protocol message should be constructed, serialized and piped to the connection. Serialization can be centralized using a special handler that takes any protocol message, encodes it into bytes and write it into the connection.

#### Actionables
- Create `ProtocolMessage` protocol message handler in `Session`. This should encode the input `ProtocolMessage`
- Create `GetPeers` actor message handler in `Session`. This should construct a `GetPeers` protocol message and forward it to the `Session` itself so that it is encoded and piped into the connection.

#### Remarks
- The `ProtocolMessage` name used in this issue description is currently called `Message` in the source code. This can be quite confusing because it may lead to confusion between protocol messages and actor messages. For the sake of clarity, we should take this opportunity to rename it to `ProtocolMessage`.

### LINKED ISSUE #96: Implement peer discovery protocol in Session actor
#### Context
- While #90 implements the `SessionsManager` side of the protocol, this implements the `Session` side.
- The peer discovery protocol allows new nodes in the network to get a more global vision of the network so that they can freely choose their peers and they don't have to trust the same nodes they used for bootstrapping.
- This strategy is very positive for descentralization, trustlessness and privacy as it makes nodes constantly change their position in the network graph.
- New peers are discovered by asking our peers for the addresses of their own peers.
- `Session` actors never initiate the peer discovery protocol by themselves. Instead, they are requested to do so by the `SessionManager`, which periodically sends `GetPeers` to a random session through `Anycast` (see #87, #89 and #90).
- Part of this functionality (adding the `GetPeers` handler to `Session`) is already tackled by issue #92.

#### Actionables
- Make `Session` react to `GetPeer` messages coming any kind of connections (both inbound and outbound) and respond immediately a `Peers` protocol message with the list of known peers (as returned by `PeersManager`'s `GetPeers` handler).

### PR DESCRIPTION
**Pull request by Tommytrg**
This PR includes:
- Implement get peers handler to session actor(#92)
- Implement peer discovery protocol in session actor(#96)

Tests for #92 and #96 will be implemented in #134 and #135.

Close #92 
Close #96

### CODE REVIEW

File: core/src/actors/session.rs [hunk 1]
```diff
@@ -125,11 +130,53 @@ impl WriteHandler<Error> for Session {}
...
+                                                // FIXME(#72): a full stop of the session is not correct (unregister should
+                                                // be skipped)
+                                                ctx.stop();
+                                            }
+                                        }
+                                        actix::fut::ok(())
+                                    })
+                                    .wait(ctx);
+                            }
+                            _ => panic!(""),
```
Comment (mmartinbar) [inline core/src/actors/session.rs]: Don't know if panicking is the best option here. I would discuss this with the team or link this as well to issue #72 , to be discussed in the near future.

File: core/src/actors/session.rs [hunk 2]
```diff
@@ -125,11 +131,21 @@ impl WriteHandler<Error> for Session {}
...
             Request::Message(message) => {
                 debug!("Session {} received message: {:?}", self.address, message);
+                let decoded_msg = WitnetMessage::try_from(message.to_vec());
+                match decoded_msg {
+                    Err(err) => debug!("Error decoding message: {:?}", err),
+                    Ok(msg) => match msg.kind {
+                        Command::GetPeers => {
+                            handle_get_peers_message(self, ctx);
+                        }
+                        _ => warn!("{:?} message not implemented", msg.kind),
```
Comment (aesedepece) [inline core/src/actors/session.rs]:
```suggestion
                        _ => warn!("Received a message of kind \"{:?}\", which is not implemented yet", msg.kind),
```

File: core/src/actors/session.rs [hunk 3]
```diff
@@ -154,5 +173,43 @@ impl Handler<GetPeers> for Session {
...
+    let peers_manager_addr = System::current()
+        .registry()
+        .get::<peers_manager::PeersManager>();
+
+    // Start chain of actions
+    peers_manager_addr
+        // Send GetPeer message to PeersManager actor
+        // This returns a Request Future, representing an asynchronous message sending process
+        .send(peers_manager::GetPeers)
+        // Convert a normal future into an ActorFuture
```
Comment (jrmoreau) [inline core/src/actors/session.rs]: Is `normal future` the same as `Request Future`?

File: core/src/actors/session.rs [hunk 4]
```diff
@@ -154,5 +173,43 @@ impl Handler<GetPeers> for Session {
 
     fn handle(&mut self, _msg: GetPeers, _: &mut Context<Self>) {
         debug!("GetPeers message should be sent through the network");
+        // Create get peers message
+        let get_peers_msg: Vec<u8> = WitnetMessage::build_get_peers().into();
+        // Write get peers message in session
```
Comment (jrmoreau) [inline core/src/actors/session.rs]:
```suggestion
        // Write GetPeers message in session
```

File: core/src/actors/session.rs [hunk 5]
```diff
@@ -154,5 +173,43 @@ impl Handler<GetPeers> for Session {
 
     fn handle(&mut self, _msg: GetPeers, _: &mut Context<Self>) {
         debug!("GetPeers message should be sent through the network");
+        // Create get peers message
```
Comment (jrmoreau) [inline core/src/actors/session.rs]:
```suggestion
        // Create GetPeers message
```

File: core/src/actors/session.rs [hunk 6]
```diff
@@ -135,14 +141,27 @@ impl WriteHandler<Error> for Session {}
 /// Implement `StreamHandler` trait in order to use `Framed` with an actor
 impl StreamHandler<Request, Error> for Session {
     /// This is main event loop for client requests
-    fn handle(&mut self, msg: Request, _ctx: &mut Self::Context) {
+    fn handle(&mut self, msg: Request, ctx: &mut Self::Context) {
         // Handle different types of requests
```
Comment (jrmoreau) [inline core/src/actors/session.rs]:
```suggestion
        // Handles different types of requests
```

File: core/src/actors/session.rs [hunk 7]
```diff
@@ -135,14 +141,27 @@ impl WriteHandler<Error> for Session {}
 /// Implement `StreamHandler` trait in order to use `Framed` with an actor
 impl StreamHandler<Request, Error> for Session {
     /// This is main event loop for client requests
```
Comment (jrmoreau) [inline core/src/actors/session.rs]:
```suggestion
    /// This is the main event loop for client requests
```

File: core/src/actors/session.rs [hunk 8]
```diff
@@ -135,14 +141,27 @@ impl WriteHandler<Error> for Session {}
 /// Implement `StreamHandler` trait in order to use `Framed` with an actor
```
Comment (jrmoreau) [inline core/src/actors/session.rs]:
```suggestion
/// This implements the `StreamHandler` trait in order to use `Framed` with an actor
```

**Pull request merged**